File: packages/firebase_analytics/ios/Classes/FirebaseAnalyticsPlugin.m
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "FirebaseAnalyticsPlugin.h"

#import "Firebase/Firebase.h"

@implementation FLTFirebaseAnalyticsPlugin {
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
  FlutterMethodChannel *channel =
      [FlutterMethodChannel methodChannelWithName:@"plugins.flutter.io/firebase_analytics"
                                  binaryMessenger:[registrar messenger]];
  FLTFirebaseAnalyticsPlugin *instance = [[FLTFirebaseAnalyticsPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (instancetype)init {
  self = [super init];
  if (self) {
    if (![FIRApp appNamed:@"__FIRAPP_DEFAULT"]) {
      NSLog(@"Configuring the default Firebase app...");
      [FIRApp configure];
      NSLog(@"Configured the default Firebase app %@.", [FIRApp defaultApp].name);
    }
  }
  return self;
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
  if ([@"logEvent" isEqualToString:call.method]) {
    NSString *eventName = call.arguments[@"name"];
    id parameterMap = call.arguments[@"parameters"];

    if (parameterMap != [NSNull null]) {
      [FIRAnalytics logEventWithName:eventName parameters:parameterMap];
    } else {
      [FIRAnalytics logEventWithName:eventName parameters:nil];
    }

    result(nil);
  } else if ([@"setUserId" isEqualToString:call.method]) {
    NSString *userId = call.arguments;
    [FIRAnalytics setUserID:userId];
    result(nil);
  } else if ([@"setCurrentScreen" isEqualToString:call.method]) {
    NSString *screenName = call.arguments[@"screenName"];
    NSString *screenClassOverride = call.arguments[@"screenClassOverride"];
    [FIRAnalytics setScreenName:screenName screenClass:screenClassOverride];
    result(nil);
  } else if ([@"setUserProperty" isEqualToString:call.method]) {
    NSString *name = call.arguments[@"name"];
    NSString *value = call.arguments[@"value"];
    [FIRAnalytics setUserPropertyString:value forName:name];
    result(nil);
  } else if ([@"setAnalyticsCollectionEnabled" isEqualToString:call.method]) {
    NSNumber *enabled = [NSNumber numberWithBool:call.arguments];
    [FIRAnalytics setAnalyticsCollectionEnabled:[enabled boolValue]];
    result(nil);
  } else if ([@"resetAnalyticsData" isEqualToString:call.method]) {
    [FIRAnalytics resetAnalyticsData];
    result(nil);
  } else {
    result(FlutterMethodNotImplemented);
  }
}

@end

File: packages/firebase_analytics/android/src/main/java/io/flutter/plugins/firebaseanalytics/FirebaseAnalyticsPlugin.java
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.firebaseanalytics;

import android.app.Activity;
import android.os.Bundle;
import com.google.firebase.FirebaseApp;
import com.google.firebase.analytics.FirebaseAnalytics;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry;
import java.util.Map;

/** Flutter plugin for Firebase Analytics. */
public class FirebaseAnalyticsPlugin implements MethodCallHandler {
  private final PluginRegistry.Registrar registrar;
  private final FirebaseAnalytics firebaseAnalytics;

  public static void registerWith(PluginRegistry.Registrar registrar) {
    final MethodChannel channel =
        new MethodChannel(registrar.messenger(), "plugins.flutter.io/firebase_analytics");
    channel.setMethodCallHandler(new FirebaseAnalyticsPlugin(registrar));
  }

  private FirebaseAnalyticsPlugin(PluginRegistry.Registrar registrar) {
    this.registrar = registrar;
    FirebaseApp.initializeApp(registrar.context());
    this.firebaseAnalytics = FirebaseAnalytics.getInstance(registrar.context());
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    switch (call.method) {
      case "logEvent":
        handleLogEvent(call, result);
        break;
      case "setUserId":
        handleSetUserId(call, result);
        break;
      case "setCurrentScreen":
        handleSetCurrentScreen(call, result);
        break;
      case "setAnalyticsCollectionEnabled":
        handleSetAnalyticsCollectionEnabled(call, result);
        break;
      case "setMinimumSessionDuration":
        handleSetMinimumSessionDuration(call, result);
        break;
      case "setSessionTimeoutDuration":
        handleSetSessionTimeoutDuration(call, result);
        break;
      case "setUserProperty":
        handleSetUserProperty(call, result);
        break;
      case "resetAnalyticsData":
        handleResetAnalyticsData(call, result);
        break;
      default:
        result.notImplemented();
        break;
    }
  }

  private void handleLogEvent(MethodCall call, Result result) {
    @SuppressWarnings("unchecked")
    Map<String, Object> arguments = (Map<String, Object>) call.arguments;
    final String eventName = (String) arguments.get("name");

    @SuppressWarnings("unchecked")
    final Bundle parameterBundle =
        createBundleFromMap((Map<String, Object>) arguments.get("parameters"));
    firebaseAnalytics.logEvent(eventName, parameterBundle);
    result.success(null);
  }

  private void handleSetUserId(MethodCall call, Result result) {
    final String id = (String) call.arguments;
    firebaseAnalytics.setUserId(id);
    result.success(null);
  }

  private void handleSetCurrentScreen(MethodCall call, Result result) {
    @SuppressWarnings("unchecked")
    Activity activity = registrar.activity();
    if (activity == null) {
      result.error("no_activity", "handleSetCurrentScreen requires a foreground activity", null);
      return;
    }
    Map<String, Object> arguments = (Map<String, Object>) call.arguments;
    final String screenName = (String) arguments.get("screenName");
    final String screenClassOverride = (String) arguments.get("screenClassOverride");

    firebaseAnalytics.setCurrentScreen(activity, screenName, screenClassOverride);
    result.success(null);
  }

  private void handleSetAnalyticsCollectionEnabled(MethodCall call, Result result) {
    final Boolean enabled = (Boolean) call.arguments;
    firebaseAnalytics.setAnalyticsCollectionEnabled(enabled);
    result.success(null);
  }

  private void handleSetMinimumSessionDuration(MethodCall call, Result result) {
    final Integer milliseconds = (Integer) call.arguments;
    firebaseAnalytics.setMinimumSessionDuration(milliseconds);
    result.success(null);
  }

  private void handleSetSessionTimeoutDuration(MethodCall call, Result result) {
    final Integer milliseconds = (Integer) call.arguments;
    firebaseAnalytics.setSessionTimeoutDuration(milliseconds);
    result.success(null);
  }

  private void handleSetUserProperty(MethodCall call, Result result) {
    @SuppressWarnings("unchecked")
    Map<String, Object> arguments = (Map<String, Object>) call.arguments;
    final String name = (String) arguments.get("name");
    final String value = (String) arguments.get("value");

    firebaseAnalytics.setUserProperty(name, value);
    result.success(null);
  }

  private void handleResetAnalyticsData(MethodCall call, Result result) {
    firebaseAnalytics.resetAnalyticsData();
    result.success(null);
  }

  private static Bundle createBundleFromMap(Map<String, Object> map) {
    if (map == null) {
      return null;
    }

    Bundle bundle = new Bundle();
    for (Map.Entry<String, Object> jsonParam : map.entrySet()) {
      final Object value = jsonParam.getValue();
      final String key = jsonParam.getKey();
      if (value instanceof String) {
        bundle.putString(key, (String) value);
      } else if (value instanceof Integer) {
        bundle.putInt(key, (Integer) value);
      } else if (value instanceof Long) {
        bundle.putLong(key, (Long) value);
      } else if (value instanceof Double) {
        bundle.putDouble(key, (Double) value);
      } else if (value instanceof Boolean) {
        bundle.putBoolean(key, (Boolean) value);
      } else {
        throw new IllegalArgumentException(
            "Unsupported value type: " + value.getClass().getCanonicalName());
      }
    }
    return bundle;
  }
}
